mbed_host_tests/__init__.py

```diff
@@ -430,91 +357,126 @@ def execute(self):
-            else:
-                self.notify("HOST: Error! No host test name defined in preamble")
-                self.print_result(self.RESULT_ERROR)
-
-            if result is not None:
-                self.print_result(result)
-            else:
-                self.notify("HOST: Passive mode...")
-        except Exception, e:
-            print str(e)
+        result = self.run_test()
+        if result == True:
+            self.print_result(self.RESULT_SUCCESS)
+        elif result == False:
+            self.print_result(self.RESULT_FAILURE)
+        elif result is None:
             self.print_result(self.RESULT_ERROR)
 
-    def run_test(self, test_supervisor):
+    def run_test(self):
         result = None
-        try:
-            self.test_supervisor = test_supervisor
-            self.test_supervisor.setup()
-            result = self.test_supervisor.test(self)    # This is blocking, waits for {end}
-        except Exception, e:
-            print str(e)
-            self.print_result(self.RESULT_ERROR)
-        finally:
+        timeout_duration = 10       # Default test case timeout
+        event_queue = Queue()       # Events from DUT to host
+        dut_event_queue = Queue()   # Events from host to DUT {k;v}
+
+        callbacks = {
+            "__notify_prn" : lambda k, v, t : self.logger.prn_inf(v)
+        }
+
+        # if True we will allow host test to consume all events after test is funished
+        callbacks_consume = True
+        self.test_supervisor = None
+
+        config = {
+            "digest" : "serial",
+            "port" : self.mbed.port,
+            "baudrate" : self.mbed.serial_baud,
+            "program_cycle_s" : self.options.program_cycle_s
+            }
+
+        self.logger.prn_inf("starting host test process...")
+        start_time = time()
+        args = (event_queue, dut_event_queue, self.prn_lock, config)
+        p = Process(target=conn_process, args=args)
+        p.deamon = True
+        p.start()
+
+        consume_preamble_events = True
+        while (time() - start_time) < timeout_duration:
+            # Handle default events like timeout, host_test_name, ...
+            if event_queue.qsize():
+                (key, value, timestamp) = event_queue.get()
+
+                if consume_preamble_events:
+                    if key == 'timeout':
+                        self.logger.prn_inf("setting timeout to: %d sec"% int(value))
+                        start_time = time()
+                        timeout_duration = int(value) # New timeout
+                    elif key == 'host_test_name':
+                        self.test_supervisor = get_host_test(value)
+                        if self.test_supervisor:
+                            self.test_supervisor.event_queue = event_queue
+                            self.test_supervisor.dut_event_queue = dut_event_queue
+                            self.test_supervisor.setup()
```

Comment:
We expect setup() sets  the callbacks, as the condition below checks if there are any callbacks. What does happen if I do:

```
# my host test
def test():
  self.register_callback();
```
Does this fail anyhow? My host test will print an error? registering callback in the test function?